### PR TITLE
Product entity in Orcid push

### PR DESCRIPTION
**Pull request by floriangantner**
## References
* Fixes https://github.com/4Science/DSpace/issues/426
* Related to 4Science/Rest7Contract#5 
* Counterpart Frontend/Angular https://github.com/4Science/dspace-angular/pull/73  

## Description
Support for Orcid Push of `Product` Entity. Allow construction of orcid work object from product entity and support for sync-mode setting change for product entity.

## Instructions for Reviewers
List of changes in this PR:
* new metadata field to hold the sync setting for product `dspace.orcid.sync-products`
* new OrcidEntityType Product and corresponding Factory `OrcidProductWorkFactory` creating orcid Work objects based on Product entity.
* new `OrcidProductWorkFieldMapping` to use some entity-specific mapping (fields might differ from publication)
  * extended NISO-Roles in Orcid (https://info.orcid.org/ufaqs/what-contributor-information-should-i-include-when-adding-works-or-funding-items/) are not map-able using the current java orcid model.
  * the mapping can be configured in the `orcid.cfg` or in the `orcid-services.xml` the same as the orcid publication factory
  * citation crosswalk to csl/bibtex is not working with Products
* simple Mapping dspace product to orcid types
  * might be improved and double checked
  * default worktype for products is `data-set`
* extended profile orcid sync preferences operation to support the change of the sync settings for the corresponding profile

* Tests: copied and aligned tests for creation of the orcid work objects for Product entity

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [ ] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

File: dspace/config/spring/api/orcid-services.xml [hunk 1]
```diff
@@ -46,6 +46,9 @@
     			<bean class="org.dspace.orcid.model.factory.impl.OrcidWorkFactory">
     				<property name="fieldMapping" ref="orcidWorkFactoryFieldMapping" />
     			</bean>
+				<bean class="org.dspace.orcid.model.factory.impl.OrcidProductWorkFactory">
+					<property name="fieldMapping" ref="orcidProductWorkFactoryFieldMapping" />
+				</bean>
```
Comment (abollini) [inline dspace/config/spring/api/orcid-services.xml:51]: can you please fix the indentation?

**Review by abollini — COMMENTED** [review]
thanks @floriangantner for this contribution. The code looks good I have just minor comments inline, can you check?
are you willing to perform the same change also to add support for patent?

File: dspace/config/spring/api/orcid-services.xml (hunk 1 above)
Comment (floriangantner) [reply dspace/config/spring/api/orcid-services.xml]: I've fixed the indentation

File: dspace-api/src/main/java/org/dspace/orcid/model/factory/impl/OrcidProductWorkFactory.java [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * Implementation of {@link OrcidEntityFactory} that creates instances of
+ * {@link Work}. Copy of {@link OrcidWorkFactory}
+ * Adapted for Product Entity with own mapping in {@link org.dspace.orcid.model.OrcidProductWorkFieldMapping}
+ *
+ * @author Luca Giamminonni (luca.giamminonni at 4science.it)
```
[the remark this answers is not in the document]
Comment (floriangantner) [reply dspace-api/src/main/java/org/dspace/orcid/model/factory/impl/OrcidProductWorkFactory.java]: I've dropped the author attribution

File: dspace-api/src/main/java/org/dspace/orcid/model/OrcidProductWorkFieldMapping.java [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+import org.dspace.util.SimpleMapConverter;
+import org.orcid.jaxb.model.common.CitationType;
+import org.orcid.jaxb.model.common.ContributorRole;
+import org.orcid.jaxb.model.v3.release.record.Work;
+
+/**
+ * Class that contains all the mapping between {@link Work} and DSpace metadata
+ * fields. Adapted for Product entity
+ *
+ * @author Luca Giamminonni (luca.giamminonni at 4science.it)
```
[the remark this answers is not in the document]
Comment (floriangantner) [reply dspace-api/src/main/java/org/dspace/orcid/model/OrcidProductWorkFieldMapping.java]: I've removed the author attribution and linked the OrcidPublicationWork

**Comment by floriangantner** [comment]
> thanks @floriangantner for this contribution. The code looks good I have just minor comments inline, can you check? are you willing to perform the same change also to add support for patent?

I've checked the comments. Support for patent will soon come in some own PR.

**Comment by olli-gold** [comment]
I can confirm, that we have integrated this PR into our production system already and it's working fine. So I guess it should be easily mergable. TORE is on version 2023.02.02.

**Pull request closed without merging**